### ISSUE TITLE
Shutdown is delayed when a sink is unhealthy

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#889]
Creating a separate issue for the discussions at https://github.com/timberio/vector/issues/874#issuecomment-532554770. The behavior described by @loony-bean dooes not appear to be ideal:

> FWIW I've observed this behaviour only when some services are not healthy

```
Sep 18 10:12:48.117 ERROR healthcheck{name=cloudwatch}: vector::topology::builder: Healthcheck: Failed Reason: deadline has elapsed
Sep 18 10:13:08.137 ERROR sink{name="cloudwatch"}: vector::sinks::util: request failed. error=Couldn't find AWS credentials in environment, credentials file, or IAM role.
Sep 18 10:13:25.946  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 54 seconds left
Sep 18 10:13:30.945  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 49 seconds left
```

Why is Vector waiting on CloudWatch at all? It appears the sink never started in the first place.

**Comment by LucioFranco** [comment]
@loony-bean do you have your config for this?

**Comment by loony-bean** [comment]
This is interesting.

So, basically I unset my AWS variables and run
```sh
$ cargo run -- --config config/examples/complete/file_to_cloudwatch_metrics.tom
```

And IFF I `^C` the program after the first error of the following kind, the program starts the count down

```
Sep 18 18:52:13.551 ERROR healthcheck{name=cloudwatch}: vector::topology::builder: Healthcheck: Failed Reason: deadline has elapsed
^CSep 18 18:52:26.527  INFO vector: Shutting down.
Sep 18 18:52:31.533  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 54 seconds left
Sep 18 18:52:36.532  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 49 seconds left
Sep 18 18:52:41.531  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 44 seconds left
Sep 18 18:52:46.531  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 39 seconds left
Sep 18 18:52:51.530  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 34 seconds left
Sep 18 18:52:56.534  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 29 seconds left
Sep 18 18:53:01.534  INFO vector::topology: Shutting down... Waiting on: cloudwatch. 24 seconds left
```

Indeed it stops a bit earlier.

Also there is no count down if I `^C` early on start up, or after few minutes of uptime.

**Comment by LucioFranco** [comment]
@loony-bean Ok, I can test this a bit later, I think I may know why but if possible could you create a gist of the log output when you run it with full trace?

`cargo run -- --config config/examples/complete/file_to_cloudwatch_metrics.tom -vvv`

**Comment by loony-bean** [comment]
@LucioFranco Sure!
https://gist.github.com/loony-bean/8a646a0aff88fb9d7d48bc4fe4556bc1

**Comment by LucioFranco** [comment]
Ok, so I think this is happening with all the rusoto sinks. It looks like the global http client that is provided with the library attempts to fetch the credentials via this provider https://docs.rs/rusoto_credential/0.17.0/rusoto_credential/struct.ChainProvider.html which attempts to fetch 4 different types of credentials. Since it can not find the credentials it has to go through all 4 providers which I want to assume is what is being slow here, not sure why that would take 15 seconds. Then it has to do this for each batch that the batched sink is holding. This may be the reason it is slow.

So we probably need to debug more what rusoto is trying to do. One possible solution to this may be to replace the default http client with our own and fetch the credentials only once at build time.

**Comment by binarylogic** [comment]
I'm confused. Why are we seeing these lines:

> Sep 18 19:25:04.752 DEBUG sink{name="cloudwatch"}: vector::sinks::aws_cloudwatch_metrics: sending data. input=...

When the sink failed to pass the healthcheck?

> Sep 18 19:25:14.723 ERROR healthcheck{name=cloudwatch}: log: Healthcheck: Failed Reason: deadline has elapsed

**Comment by LucioFranco** [comment]
@binarylogic I believe the require healthchecks option is not on and therefore we don't wait for the healthchecks to complete. Now that I think about it we probably want the opposite behavior?

**Comment by binarylogic** [comment]
@ktff it's possible this is no longer an issue with the shutdown changes but I want to verify.

**Comment by ktff** [comment]
@binarylogic CloudWatch sink now properly shutdowns, when asked to. So I'll close this issue.
